keel-web/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -0,0 +1,169 @@
+package com.netflix.spinnaker.keel.services
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.core.api.ArtifactSummary
+import com.netflix.spinnaker.keel.core.api.ArtifactSummaryInEnvironment
+import com.netflix.spinnaker.keel.core.api.ArtifactVersionSummary
+import com.netflix.spinnaker.keel.core.api.ArtifactVersions
+import com.netflix.spinnaker.keel.core.api.EnvironmentSummary
+import com.netflix.spinnaker.keel.pause.ActuationPauser
+import com.netflix.spinnaker.keel.persistence.ArtifactRepository
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.persistence.ResourceStatus
+import com.netflix.spinnaker.keel.persistence.ResourceSummary
+import com.netflix.spinnaker.kork.web.exceptions.InvalidRequestException
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+/**
+ * Service object that offers high-level APIs for application-related operations.
+ */
+@Component
+class ApplicationService(
+  private val pauser: ActuationPauser,
+  private val repository: KeelRepository,
+  private val artifactRepository: ArtifactRepository
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  fun hasManagedResources(application: String) = repository.hasManagedResources(application)
+
+  /**
+   * Returns a list of [ResourceSummary] for the specified application.
+   *
+   * This function assumes there's a single delivery config associated with the application.
+   */
+
+  fun getResourceSummariesFor(application: String): List<ResourceSummary> {
+    var resources = repository.getSummaryByApplication(application)
+    resources = resources.map { summary ->
+      if (pauser.resourceIsPaused(summary.id)) {
+        // we only update the status if the individual resource is paused,
+        // because the application pause is reflected in the response as a top level key.
+        summary.copy(status = ResourceStatus.PAUSED)
+      } else {
+        summary
+      }
+    }
+    // val constraintStates = repository.constraintStateFor(application)
+
+    return resources
+  }
+
+  fun getEnvironmentSummariesFor(application: String): List<EnvironmentSummary> =
+    getFirstDeliveryConfigFor(application)
+      ?.let { deliveryConfig ->
+        artifactRepository.getEnvironmentSummaries(deliveryConfig)
+      }
+      ?: emptyList()
+
+  /**
+   * Returns a list of [ArtifactSummary] for the specified application by traversing the list of [EnvironmentSummary]
+   * for the application and reindexing the data so that it matches the right format.
+   *
+   * This function assumes there's a single delivery config associated with the application.
+   */
+  fun getArtifactSummariesFor(application: String): List<ArtifactSummary> {
```

Comment:
This function is fairly long and involved, but I did not want (frankly, I did not feel skilled enough) to write a SQL query like @robfletcher wrote to get the environment summaries above. Also, 99% of the data was already present in the environment summaries, so I took the approach of transforming the data to fit the format here. I added comments throughout the function -- hopefully they help.

Reply:
In general, I prefer a longer function comment at the top explaining the steps, then well-named variables (as you've done) to make it clear to read. For example,
```kotlin
// map of version summaries by artifact
    val versionSummariesByArtifact = ...
```
is a comment that doesn't add much IMO, because the variable has a great name.

Reply:
Yeah, good point. I'll reorganize the comments.